File: src/algorithms/astar.py
import heapq
import math
from .base import PathfindingAlgorithm


# ── Shared A* search engine ───────────────────────────────────────────────────

def _astar(grid, start: tuple, end: tuple, heuristic) -> tuple:
    g       = {start: 0}
    prev    = {start: None}
    closed  = set()
    visited = []
    counter = 0
    open_set = []
    heapq.heappush(open_set, (heuristic(start, end), counter, start))

    while open_set:
        _, _, node = heapq.heappop(open_set)
        if node in closed:
            continue
        closed.add(node)
        visited.append(node)

        if node == end:
            break

        for nbr, edge_cost in grid.get_neighbors_with_cost(node):
            if nbr in closed:
                continue
            new_g = g[node] + edge_cost
            if nbr not in g or new_g < g[nbr]:
                g[nbr]    = new_g
                prev[nbr] = node
                counter  += 1
                heapq.heappush(open_set, (new_g + heuristic(nbr, end), counter, nbr))

    # Reconstruct
    path, n = [], end
    while n is not None:
        path.append(n)
        n = prev.get(n)
    path.reverse()
    if not path or path[0] != start:
        path = []
    return path, visited, g


# ── Heuristics ────────────────────────────────────────────────────────────────

def _manhattan(a, b):
    return abs(a[0] - b[0]) + abs(a[1] - b[1])

def _euclidean(a, b):
    return math.hypot(a[0] - b[0], a[1] - b[1])

def _hex_cube_dist(a, b):
    """Cube-coordinate hex distance (converted from odd-r offset)."""
    def to_cube(pos):
        c, r = pos
        x = c - (r - (r & 1)) // 2
        z = r
        return x, -x - z, z
    ax, ay, az = to_cube(a)
    bx, by, bz = to_cube(b)
    return max(abs(ax - bx), abs(ay - by), abs(az - bz))


# ── Public algorithm classes ──────────────────────────────────────────────────

class AStarManhattan(PathfindingAlgorithm):
    """A* with Manhattan distance — best for square 4-way grids."""

    @property
    def name(self):  return "A* Manhattan"
    @property
    def color(self): return (255, 220, 30)

    def find_path(self, grid, start, end):
        return _astar(grid, start, end, _manhattan)


class AStarEuclidean(PathfindingAlgorithm):
    """A* with Euclidean distance — works on any geometry."""

    @property
    def name(self):  return "A* Euclidean"
    @property
    def color(self): return (255, 80, 80)

    def find_path(self, grid, start, end):
        return _astar(grid, start, end, _euclidean)


class AStarHex(PathfindingAlgorithm):
    """A* with cube-coordinate hex distance — optimal heuristic for hex grids."""

    @property
    def name(self):  return "A* Hex (cube dist)"
    @property
    def color(self): return (80, 255, 150)

    def find_path(self, grid, start, end):
        return _astar(grid, start, end, _hex_cube_dist)


# Re-export base class so consumers can do `from algorithms.astar import PathfindingAlgorithm`
from .base import PathfindingAlgorithm  # noqa: F401,E402

File: web/src/algorithms/astar.js
// Mirrors src/algorithms/astar.py — shared A* engine + three heuristic classes.

import { PathfindingAlgorithm } from './base.js';
import { MinHeap } from '../utils.js';

// ── Shared A* engine ──────────────────────────────────────────────────────────

function astar(grid, start, end, heuristic) {
    const g       = new Map([[start, 0]]);
    const prev    = new Map([[start, null]]);
    const closed  = new Set();
    const visited = [];
    let   counter = 0;
    const open    = new MinHeap();
    open.push([heuristic(start, end), counter, start]);

    while (open.size > 0) {
        const [, , node] = open.pop();
        if (closed.has(node)) continue;
        closed.add(node);
        visited.push(node);

        if (node === end) break;

        for (const [nbr, edgeCost] of grid.getNeighborsWithCost(node)) {
            if (closed.has(nbr)) continue;
            const newG = g.get(node) + edgeCost;
            if (!g.has(nbr) || newG < g.get(nbr)) {
                g.set(nbr, newG);
                prev.set(nbr, node);
                counter++;
                open.push([newG + heuristic(nbr, end), counter, nbr]);
            }
        }
    }

    // Reconstruct path
    const path = [];
    let n = end;
    while (n !== null && n !== undefined) {
        path.push(n);
        n = prev.get(n);
    }
    path.reverse();

    return {
        path:         (path.length > 0 && path[0] === start) ? path : [],
        visitedOrder: visited,
        costMap:      g,
    };
}

// ── Heuristics ────────────────────────────────────────────────────────────────
// All heuristics accept posKey strings ("c,r") and parse them internally.

function parseKey(key) {
    const i = key.indexOf(',');
    return [parseInt(key, 10), parseInt(key.slice(i + 1), 10)];
}

function manhattan(a, b) {
    const [ac, ar] = parseKey(a), [bc, br] = parseKey(b);
    return Math.abs(ac - bc) + Math.abs(ar - br);
}

function euclidean(a, b) {
    const [ac, ar] = parseKey(a), [bc, br] = parseKey(b);
    return Math.hypot(ac - bc, ar - br);
}

function hexCubeDist(a, b) {
    // Convert odd-r offset → cube coordinates, then Chebyshev distance
    function toCube(key) {
        const [c, r] = parseKey(key);
        const x = c - ((r - (r & 1)) >> 1);
        const z = r;
        return [x, -x - z, z];
    }
    const [ax, ay, az] = toCube(a);
    const [bx, by, bz] = toCube(b);
    return Math.max(Math.abs(ax - bx), Math.abs(ay - by), Math.abs(az - bz));
}

// ── Public algorithm classes ──────────────────────────────────────────────────

/** A* with Manhattan distance — optimal for square 4-way grids. */
export class AStarManhattan extends PathfindingAlgorithm {
    get name()  { return 'A* Manhattan'; }
    get color() { return [255, 220, 30]; }
    findPath(grid, start, end) { return astar(grid, start, end, manhattan); }
}

/** A* with Euclidean distance — works on any grid geometry. */
export class AStarEuclidean extends PathfindingAlgorithm {
    get name()  { return 'A* Euclidean'; }
    get color() { return [255, 80, 80]; }
    findPath(grid, start, end) { return astar(grid, start, end, euclidean); }
}

/** A* with cube-coordinate hex distance — optimal heuristic for hex grids. */
export class AStarHex extends PathfindingAlgorithm {
    get name()  { return 'A* Hex (cube dist)'; }
    get color() { return [80, 255, 150]; }
    findPath(grid, start, end) { return astar(grid, start, end, hexCubeDist); }
}
